### PR TITLE
axolotl example

### PR DESCRIPTION
**Pull request by manishiitg**
Added example for using axolotl

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: llm/axolotl/mistral/qlora-checkpoint.yaml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+datasets:
+  - path: mhenrichsen/alpaca_2k_test
+    type: alpaca
+
+dataset_prepared_path: /sky-notebook/alpaca_2k_test/last_run_prepared
+val_set_size: 0.05
+output_dir: /sky-notebook/alpaca_2k_test
+
+hub_model_id: manishiitg/mistral-alpaca_2k_test # TODO: Replace with hub model id
+hf_use_auth_token: false # TODO: push as private or public model
```
Comment (concretevitamin) [inline llm/axolotl/mistral/qlora-checkpoint.yaml:19]: See the other comment - should comment out these?

File: llm/axolotl/axolotl-spot.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
+name: axolotl
```
Comment (concretevitamin) [inline llm/axolotl/axolotl-spot.yaml:1]:
```suggestion
# Usage:
#
#  Unmanaged spot (no auto-recovery; for debugging):
#   HF_TOKEN=abc BUCKET=<unique-name> sky launch -c axolotl-spot axolotl-spot.yaml --env HF_TOKEN --env BUCKET -i30 --down
#
#  Managed spot (auto-recovery; for full runs):
#   HF_TOKEN=abc BUCKET=<unique-name> sky spot launch -n axolotl-spot axolotl-spot.yaml --env HF_TOKEN --env BUCKET

name: axolotl
```

File: llm/axolotl/axolotl-spot.yaml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+resources:
+  accelerators: A100:1
+  cloud: gcp # optional
```
Comment (concretevitamin) [inline llm/axolotl/axolotl-spot.yaml:5]:
```suggestion
  cloud: gcp # optional
  use_spot: True
```

File: llm/axolotl/axolotl-spot.yaml [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+workdir: mistral
+
+file_mounts:
+  /sky-notebook:
+    source: gs://sky-notebook
```
Comment (concretevitamin) [inline llm/axolotl/axolotl-spot.yaml:11]:
```suggestion
    name: ${BUCKET}
```

File: llm/axolotl/axolotl-spot.yaml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  
+  docker run --gpus all \
+    -v ~/sky_workdir:/sky_workdir \
+    -v /root/.cache:/root/.cache \
+    -v /sky-notebook:/sky-notebook \
+    winglian/axolotl:main-py3.10-cu118-2.0.1 \
+    accelerate launch -m axolotl.cli.train /sky_workdir/qlora-checkpoint.yaml
+
+envs:
+  HF_TOKEN: <your-huggingface-token> # TODO: Replace with huggingface token
```
Comment (concretevitamin) [inline llm/axolotl/axolotl-spot.yaml:32]:
```suggestion
  HF_TOKEN: <your-huggingface-token> # TODO: Replace with huggingface token
  BUCKET: <a-unique-bucket-name-to-use>
```

File: llm/axolotl/axolotl.yaml [hunk 6]
```diff
@@ -0,0 +1,32 @@
+name: axolotl
```
Comment (concretevitamin) [inline llm/axolotl/axolotl.yaml:1]:
```suggestion
# Usage:
#   HF_TOKEN=abc sky launch -c axolotl axolotl.yaml --env HF_TOKEN -y -i30 --down

name: axolotl
```

File: llm/axolotl/readme.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
+This example is using the same config located https://github.com/OpenAccess-AI-Collective/axolotl/tree/main/examples/mistral
+
+
+
+Simple training example
+```
+sky launch -c axolotl axolotl.yaml -y
```
Comment (concretevitamin) [inline llm/axolotl/readme.md:7]:
```suggestion
HF_TOKEN=abc sky launch -c axolotl axolotl.yaml --env HF_TOKEN -y -i30 --down
```

File: llm/axolotl/readme.md [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+ssh -L 8888:localhost:8888 axolotl
+sky down axolotl -y
+```
+
+
+
+If you want to add use_spot: True in resources section. this won't support auto resume
```
Comment (concretevitamin) [inline llm/axolotl/readme.md:14]:
```suggestion
To launch an unmanaged spot instance (no auto-recovery; good for debugging)
```

File: llm/axolotl/readme.md [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+```
+sky launch -c axolotl-spot axolotl-spot.yaml -y
```
Comment (concretevitamin) [inline llm/axolotl/readme.md:16]:
```suggestion
HF_TOKEN=abc BUCKET=<unique-name> sky launch -c axolotl-spot axolotl-spot.yaml --env HF_TOKEN --env BUCKET -i30 --down
```

File: llm/axolotl/readme.md [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+ssh -L 8888:localhost:8888 axolotl-spot
+```
+
+
+Launch spot instances
```
Comment (concretevitamin) [inline llm/axolotl/readme.md:21]:
```suggestion
Launch managed spot instances (auto-recovery; for full runs):
```

File: llm/axolotl/readme.md [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+```
+sky spot launch -n axolotl-spot axolotl-spot.yaml -y
+
```
Comment (concretevitamin) [inline llm/axolotl/readme.md:24]:
```suggestion
HF_TOKEN=abc BUCKET=<unique-name> sky spot launch -n axolotl-spot axolotl-spot.yaml --env HF_TOKEN --env BUCKET
```

File: llm/axolotl/mistral/qlora.yaml [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+strict: false
+
+datasets:
+  - path: mhenrichsen/alpaca_2k_test
+    type: alpaca
+dataset_prepared_path: last_run_prepared
+val_set_size: 0.05
+output_dir: ./qlora-out
+
+hub_model_id: manishiitg/mhenrichsen-alpaca_2k_test # TODO: Replace with hub model id
```
Comment (concretevitamin) [inline llm/axolotl/mistral/qlora.yaml:17]:
```suggestion
# hub_model_id: manishiitg/mhenrichsen-alpaca_2k_test # TODO: Replace with hub model id
```

File: llm/axolotl/mistral/qlora-checkpoint.yaml [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+datasets:
+  - path: mhenrichsen/alpaca_2k_test
+    type: alpaca
+
+dataset_prepared_path: /sky-notebook/alpaca_2k_test/last_run_prepared
+val_set_size: 0.05
+output_dir: /sky-notebook/alpaca_2k_test
+
+hub_model_id: manishiitg/mistral-alpaca_2k_test # TODO: Replace with hub model id
```
Comment (concretevitamin) [inline llm/axolotl/mistral/qlora-checkpoint.yaml:18]:
```suggestion
# hub_model_id: manishiitg/mistral-alpaca_2k_test # TODO: Replace with hub model id
```

**Review by concretevitamin — APPROVED** [review]
Wow, this is super awesome @manishiitg! Tested it out and both YAMLs work great. Thanks a bunch for sharing.

Made some suggestions, mostly about passing secrets via env vars.

File: llm/axolotl/mistral/qlora.yaml (hunk 12 above)
[the remark this answers is not in the document]
Comment (manishiitg) [reply llm/axolotl/mistral/qlora.yaml]: commented this

**Comment by manishiitg** [comment]
@concretevitamin incorporated comments

**Pull request merged**